platform/ios/src/MGLMapboxEvents.m

```diff
@@ -314,38 +315,36 @@ - (void)flush {
 }
 
 - (void)pushTurnstileEvent {
-    __weak MGLMapboxEvents *weakSelf = self;
+    if (self.turnstileSendDate && [[NSDate date] timeIntervalSinceDate:self.turnstileSendDate] < 0) {
+        return;
+    }
     
-    static dispatch_once_t onceToken;
-    dispatch_once(&onceToken, ^{
-        MGLMapboxEvents *strongSelf = weakSelf;
-        
-        if (!strongSelf) {
-            return;
-        }
-        
-        NSString *vendorID = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
-        if (!vendorID) {
-            return;
-        }
-        
-        NSDictionary *turnstileEventAttributes = @{MGLEventKeyEvent: MGLEventTypeAppUserTurnstile,
-                                                   MGLEventKeyCreated: [strongSelf.rfc3339DateFormatter stringFromDate:[NSDate date]],
-                                                   MGLEventKeyVendorID: vendorID,
-                                                   MGLEventKeyEnabledTelemetry: @([[strongSelf class] isEnabled])};
-        
-        if ([MGLAccountManager accessToken] == nil) {
+    NSString *vendorID = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
+    if (!vendorID) {
+        return;
+    }
+    
+    NSDictionary *turnstileEventAttributes = @{MGLEventKeyEvent: MGLEventTypeAppUserTurnstile,
+                                               MGLEventKeyCreated: [self.rfc3339DateFormatter stringFromDate:[NSDate date]],
+                                               MGLEventKeyVendorID: vendorID,
+                                               MGLEventKeyEnabledTelemetry: @([[self class] isEnabled])};
+    
+    if ([MGLAccountManager accessToken] == nil) {
+        return;
+    }
+    
+    __weak __typeof__(self) weakSelf = self;
+    [self.apiClient postEvent:turnstileEventAttributes completionHandler:^(NSError * _Nullable error) {
+        __strong __typeof__(weakSelf) strongSelf = weakSelf;
+        if (error) {
+            [strongSelf pushDebugEvent:MGLEventTypeLocalDebug withAttributes:@{MGLEventKeyLocalDebugDescription: @"Network error",
+                                                                         @"error": error}];
             return;
         }
-        [strongSelf.apiClient postEvent:turnstileEventAttributes completionHandler:^(NSError * _Nullable error) {
-            if (error) {
-                [strongSelf pushDebugEvent:MGLEventTypeLocalDebug withAttributes:@{MGLEventKeyLocalDebugDescription: @"Network error",
-                                                                                   @"error": error}];
-                return;
-            }
-            [strongSelf writeEventToLocalDebugLog:turnstileEventAttributes];
-        }];
-    });
+        [strongSelf writeEventToLocalDebugLog:turnstileEventAttributes];
+        NSTimeInterval twentyFourHourTimeInterval = 24 * 3600;
+        strongSelf.turnstileSendDate = [[NSDate date] dateByAddingTimeInterval:twentyFourHourTimeInterval];
+    }];
```

Comment:
Note: we **will** not consider a turnstile event to have been sent in a session unless the server responds favorably.